scanpy/preprocessing/_utils.py

```diff
@@ -15,6 +18,52 @@ def _get_mean_var(X, *, axis=0):
     return mean, var
 
 
+def _pca_with_sparse(X, npcs, solver='arpack', mu=None):
```

Comment:
This should get a `random_state` argument. Make sure it will work if either a `RandomState` or `int` is passed to `sc.pp.pca`

Reply:
I don't think [scipy.sparse.linalg.svds](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html) accepts a random state.

Reply:
It would probably be easiest to normalize the random state with:

```python
random_state = sklearn.utils.check_random_state(random_state)
```

during argument handling for `sc.pp.pca` then setting `np.random.set_state(random_state.get_state())` in `_pca_with_sparse`.

Reply:
Gotcha!

Reply:
Ah, no you're right. I was getting the same result when I set the seed due to conversion from float64 -> float32 which made values similar.

`svds` does use randomness to initialize though, it would be nice if we could set that.

Reply:
It seems that setting the random state with numpy does not affect the output of `svds`? Using the same seed, with `float32` data (avg residual is ~5e-4) and with `float64`, the average residual is ~1e-14. But with different seeds, the residuals are the same in both cases.

Reply:
I think the `v0` parameter in `svds` might be what we're looking for? We could set the state with numpy and then generate a random vector to input into v0. The default option for `v0` is random, anyway.

Reply:
Found the scipy issue for this: https://github.com/scipy/scipy/issues/9023

@falexwolf, am I wrong in thinking we changed the default to arpack for reproducibility?

Reply:
According to the linked discussion, it seems that it is not as easy as setting `v0` to a random vector as the distribution you use to generate the random vector can introduce bias.

Reply:
Here's what sklearn's PCA does: https://github.com/scikit-learn/scikit-learn/blob/244d118ef77a513b487b95b721179c395cbc1660/sklearn/decomposition/_pca.py#L524-L525

They generate `v0`. This is probably the way to go, but I'm a little torn on whether to use this or get a fix upstreamed to scipy. I'm also not sure how well setting v0 like this will go for non-arpack solvers.

Reply:
> According to the linked discussion, it seems that it is not as easy as setting v0 to a random vector as the distribution you use to generate the random vector can introduce bias.

Yeah, this is what makes me unsure.

Reply:
Well, we can use the average residual between `_pca_with_sparse` (with and without random seeding) and `sklearn.decomposition.PCA` as a check to see if we are introducing bias. In my hands, using `v0` vs not makes no difference in terms of the residual.

Reply:
Also, `lobpcg` doesn't even use `v0`.

Reply:
Doesn't it? From `svds`:

```python
    if solver == 'lobpcg':

        if k == 1 and v0 is not None:
            X = np.reshape(v0, (-1, 1))
        else:
            X = np.random.RandomState(52).randn(min(A.shape), k)

        eigvals, eigvec = lobpcg(XH_X, X, tol=tol ** 2, maxiter=maxiter,
                                 largest=largest)
```

It just hard codes the random state...

Reply:
Oh, right, I missed that. Okay, so it uses `v0` when you compute a single singular vector, lol.

Reply:
> @falexwolf, am I wrong in thinking we changed the default to arpack for reproducibility?

I believe so as well. I had issues reproducing my `single-cell-tutorial` workflow on different systems without using `svd_solver='arpack'`. It still wasn't 100% reproducible afterwards, but much more similar betweem runs. Alex recommended this option at the time... and I believe the defaults were changed as a result of that as well.

Reply:
In scikit-learn==0.22.1, `TruncatedSVD` does not use the `random_state` parameter if the solver is `arpack`, which is why you still weren't getting 100% reproducibility.
```python
        X = check_array(X, accept_sparse=['csr', 'csc'],
                        ensure_min_features=2)
        random_state = check_random_state(self.random_state)

        if self.algorithm == "arpack":
            U, Sigma, VT = svds(X, k=self.n_components, tol=self.tol)
            # svds doesn't abide by scipy.linalg.svd/randomized_svd
            # conventions, so reverse its outputs.
            Sigma = Sigma[::-1]
            U, VT = svd_flip(U[:, ::-1], VT[::-1])
```